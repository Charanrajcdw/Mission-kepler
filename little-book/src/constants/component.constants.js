export const MODAL = {
  members: "Members",
  newBlog: "Add New Blog",
  remove: "Remove",
};

export const MEMBERS_LIST = {
  title: "Members",
  noMembers: "No members available",
};

export const BLOG_FORM = {
  title: "Add New Blog",
  button: "ADD",
  titlePlaceholder: "Name your blog",
  detailsPlaceholder: "Write Content Here ..",
  invalidValues: "Invalid title or description",
  invalidBlog: "Blog title already exists",
  blogAdded: "New Blog added successfully",
  localBlog: "local",
};

export const WARNING_MODAL = {
  title: "Unsaved changes will be lost. Are you sure to continue?",
  continue: "Continue",
  cancel: "Cancel",
};
