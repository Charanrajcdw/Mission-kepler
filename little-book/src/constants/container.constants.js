export const SIDEBAR = {
  logo: "Little Book",
  filter: "FILTER",
  noFilter: "Add blogs to show filters",
  viewMembers: "View Members",
  switchToDark: "Switch to Dark Mode",
  switchToLight: "Switch to Light Mode",
  blogs: "Blogs",
};
