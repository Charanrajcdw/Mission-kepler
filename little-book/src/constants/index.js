export * from "./url.constants";
export * from "./theme.constants";
