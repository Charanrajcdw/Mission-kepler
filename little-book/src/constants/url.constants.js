export const API_URLS = {
  blogs: "https://jsonmockserver.vercel.app/api/blogs",
  users: "https://jsonmockserver.vercel.app/api/users",
  images: "https://jsonmockserver.vercel.app/",
};
