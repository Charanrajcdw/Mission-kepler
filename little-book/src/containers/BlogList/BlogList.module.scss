.blogListContainer {
  background-color: var(--blog-list-bg);
  width: 33.75%;
  padding: 36px 28px 36px 40px;

  .searchBar {
    display: flex;
    flex-flow: row nowrap;
    justify-content: space-between;

    .search {
      background-color: var(--blog-input-bg);
      color: var(--blog-placeholder-bg);
      border: none;
      flex-basis: 22.5rem;
      padding: 17.5px;
      border-radius: 28px;
      font-size: 0.9rem;
      font-weight: 300;

      &::placeholder {
        color: var(--blog-placeholder-bg);
      }

      &:focus {
        outline: 1px solid var(--pink);
      }

      &:disabled {
        cursor: not-allowed;
      }
    }
  }

  .blogsContainer {
    display: flex;
    flex-direction: column;
    gap: 20px;
    height: 86vh;
    overflow: auto;
    margin: 24px 0;

    &::-webkit-scrollbar {
      display: none;
    }

    .noBlogs {
      color: var(--primary-font-bg);
    }
  }
}
