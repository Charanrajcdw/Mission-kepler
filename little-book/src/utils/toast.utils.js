import { toast } from "react-toastify";

export const showSuccessToast = (message, currentTheme) => {
  toast.success(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};

export const showWarningToast = (message, currentTheme) => {
  toast.warn(message, {
    position: "top-right",
    autoClose: 2000,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    theme: currentTheme,
  });
};
