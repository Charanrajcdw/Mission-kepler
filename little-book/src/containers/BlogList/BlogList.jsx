import { useCallback, useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import PropTypes from "prop-types";
import styles from "./BlogList.module.scss";
import { Button, BlogCard, Loader } from "../../components";
import { BLOG_LIST, ENTER, MODAL } from "../../constants";
import { getBlogs } from "../../services/blog.services";
import { blogActions } from "../../store";

const BlogList = ({ modalHandler }) => {
  const { filteredBlogData: blogs, isLoaded, currentBlog, searchTerm, isEditing } = useSelector((state) => state.blogs);
  const dispatch = useDispatch();
  const blogsRef = useRef();

  useEffect(() => {
    getBlogs().then((blogs) => dispatch(blogActions.modifyBlogs(blogs)));
  }, [dispatch]);

  useEffect(() => {
    blogsRef.current.scrollTo(0, 0);
  }, [blogs]);

  // handler function to handle search term change submits
  const submitHandler = (e) => {
    if (e.key === ENTER) dispatch(blogActions.modifySearchTerm(e.target.value));
  };

  // handler function for onClick on new blog button
  const newBtnHandler = () => {
    modalHandler(MODAL.newBlog, false);
    dispatch(blogActions.modifyEditStatus(true));
  };

  // handler function for onClick on each blog to change current blog
  const blogChangeHandler = useCallback(
    (blog) => {
      if (isEditing) {
        modalHandler(false, blogActions.modifyCurrentBlog(blog));
        return;
      }
      dispatch(blogActions.modifyCurrentBlog(blog));
    },
    [dispatch, isEditing, modalHandler]
  );

  const getBlogsContent = () => {
    let blogContent = "";
    if (blogs.length > 0) {
      blogContent = blogs.map((blog) => (
        <BlogCard key={blog.title} blog={blog} isSelected={currentBlog.title === blog.title} blogChangeHandler={blogChangeHandler} />
      ));
    } else {
      blogContent = <p className={styles.noBlogs}>{BLOG_LIST.noBlogs}</p>;
    }
    return blogContent;
  };

  return (
    <div className={styles.blogListContainer}>
      <div className={styles.searchBar}>
        <input
          type="text"
          className={styles.search}
          placeholder={BLOG_LIST.searchText}
          defaultValue={searchTerm}
          onKeyDown={submitHandler}
          disabled={isEditing}
        />
        <Button className="pinkBtn" clickHandler={newBtnHandler} disabled={isEditing}>
          {BLOG_LIST.buttonText}
        </Button>
      </div>
      <div className={styles.blogsContainer} ref={blogsRef}>
        {isLoaded ? getBlogsContent() : <Loader />}
      </div>
    </div>
  );
};

BlogList.propTypes = {
  modalHandler: PropTypes.func.isRequired,
};

export default BlogList;
