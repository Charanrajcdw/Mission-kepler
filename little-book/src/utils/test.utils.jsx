import { configureStore } from "@reduxjs/toolkit";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import ThemeContextProvider from "../contexts/theme.context";
import blogSlice from "../store/slices/blogSlice";
import { THEME } from "../constants";

const customRender = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
    ...options
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return render(ui, { wrapper: ProviderWrapper, ...options });
};

const customCreateSnapshot = (
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: { blogs: blogSlice.reducer },
      preloadedState,
    }),
    theme = THEME.light,
  } = {}
) => {
  const ProviderWrapper = ({ children }) => {
    return (
      <ThemeContextProvider theme={theme}>
        <Provider store={store}>{children}</Provider>
      </ThemeContextProvider>
    );
  };
  return renderer.create(<ProviderWrapper>{ui}</ProviderWrapper>).toJSON();
};

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
export { customRender as render, customCreateSnapshot as createSnapshot };
