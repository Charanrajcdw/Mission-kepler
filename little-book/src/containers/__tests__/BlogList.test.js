import { render, screen, cleanup, waitFor, waitForElementToBeRemoved, userEvent } from "../../utils/test.utils";
import axiosMock from "axios";
import BlogList from "../BlogList/BlogList";
import { BLOG_LIST, MODAL } from "../../constants";

jest.mock("../../components/BlogCard/BlogCard", () => ({ blog, blogChangeHandler, isSelected }) => (
  <div data-testid="blogCard" onClick={() => blogChangeHandler(blog)} data-selected={isSelected} />
));
jest.mock("../../components/Loader/Loader", () => () => <div data-testid="loader" />);
Element.prototype.scrollTo = () => {};
afterEach(cleanup);

describe("Blog List UI", () => {
  it("should render blog list with blogs", async () => {
    axiosMock.get.mockResolvedValueOnce({
      data: [
        { title: "title1", type: "type1" },
        { title: "title2", type: "type2" },
      ],
    });
    render(<BlogList modalHandler={() => {}} />);
    const searchInput = screen.getByPlaceholderText(BLOG_LIST.searchText);
    const newBlogButton = screen.getByText(BLOG_LIST.buttonText);
    expect(searchInput).toBeInTheDocument();
    expect(newBlogButton).toBeInTheDocument();
    await waitFor(() => expect(screen.getByTestId("loader")).toBeInTheDocument());
    await waitFor(() => expect(screen.queryByTestId("loader")).toBeNull());
    const membersElements = await screen.findAllByTestId("blogCard");
    expect(membersElements).toHaveLength(2);
  });

  it("should render no blogs content with no blogs", async () => {
    axiosMock.get.mockResolvedValueOnce({ data: [] });
    render(<BlogList modalHandler={() => {}} />);
    const searchInput = screen.getByPlaceholderText(BLOG_LIST.searchText);
    const newBlogButton = screen.getByText(BLOG_LIST.buttonText);
    expect(searchInput).toBeInTheDocument();
    expect(newBlogButton).toBeInTheDocument();
    await waitFor(() => expect(screen.getByTestId("loader")).toBeInTheDocument());
    await waitFor(() => expect(screen.queryByTestId("loader")).toBeNull());
    const noBlogsElement = await screen.findByText(BLOG_LIST.noBlogs);
    expect(noBlogsElement).toBeInTheDocument();
  });
});

describe("Blog List UI", () => {
  it("should display the cards that match search text", async () => {
    const user = userEvent.setup();
    axiosMock.get.mockResolvedValueOnce({
      data: [
        { title: "title1", type: "type1" },
        { title: "title2", type: "type2" },
      ],
    });
    render(<BlogList modalHandler={() => {}} />);
    const searchInput = screen.getByPlaceholderText(BLOG_LIST.searchText);
    // search for title
    await user.type(searchInput, "title");
    let blogElements;
    blogElements = await screen.findAllByTestId("blogCard");
    expect(blogElements).toHaveLength(2);
    // search for title1
    await user.type(searchInput, "1");
    blogElements = await screen.findAllByTestId("blogCard");
    expect(blogElements).toHaveLength(1);
  });

  it("should call new blog modal handler when new button is clicked", async () => {
    const user = userEvent.setup();
    const modalFn = jest.fn();
    axiosMock.get.mockResolvedValueOnce({ data: [] });
    render(<BlogList modalHandler={modalFn} />);
    const newBlogButton = screen.getByText(BLOG_LIST.buttonText);
    await user.click(newBlogButton);
    expect(modalFn).toHaveBeenCalledWith(MODAL.newBlog, false);
  });

  it("should call warning modal on blog card click if in editing state", async () => {
    const user = userEvent.setup();
    const modalFn = jest.fn();
    axiosMock.get.mockResolvedValueOnce({
      data: [
        { title: "title1", type: "type1" },
        { title: "title2", type: "type2" },
      ],
    });
    const initialBlogsData = { isEditing: true };
    render(<BlogList modalHandler={modalFn} />, { preloadedState: { blogs: initialBlogsData } });
    const blogElements = await screen.findAllByTestId("blogCard");
    await user.click(blogElements[1]);
    expect(modalFn).toHaveBeenCalledWith(false, true);
  });

  it("should call current blog update handler if not in editing state", async () => {
    const user = userEvent.setup();
    axiosMock.get.mockResolvedValueOnce({
      data: [
        { title: "title1", type: "type1" },
        { title: "title2", type: "type2" },
      ],
    });
    render(<BlogList modalHandler={() => {}} />);
    const blogElements = await screen.findAllByTestId("blogCard");
    expect(blogElements[0]).toHaveAttribute("data-selected", "true");
    expect(blogElements[1]).toHaveAttribute("data-selected", "false");
    await user.click(blogElements[1]);
    expect(blogElements[0]).toHaveAttribute("data-selected", "false");
    expect(blogElements[1]).toHaveAttribute("data-selected", "true");
  });
});

describe("Blog list snapshots", () => {
  it("should match blog list with blogs", async () => {
    axiosMock.get.mockResolvedValueOnce({
      data: [
        { title: "title1", type: "type1" },
        { title: "title2", type: "type2" },
      ],
    });
    const { asFragment } = render(<BlogList modalHandler={() => {}} />);
    await waitForElementToBeRemoved(screen.getByTestId("loader"));
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match blog list with no blogs", async () => {
    axiosMock.get.mockResolvedValueOnce({
      data: [],
    });
    const { asFragment } = render(<BlogList modalHandler={() => {}} />);
    await waitForElementToBeRemoved(screen.getByTestId("loader"));
    expect(asFragment()).toMatchSnapshot();
  });
});
