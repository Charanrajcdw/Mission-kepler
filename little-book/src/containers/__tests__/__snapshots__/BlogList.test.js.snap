// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog list snapshots should match blog list with blogs 1`] = `
<DocumentFragment>
  <div
    class="blogListContainer"
  >
    <div
      class="searchBar"
    >
      <input
        class="search"
        placeholder="Search Blogs"
        type="text"
        value=""
      />
      <button
        class="btn pinkBtn"
      >
        NEW
      </button>
    </div>
    <div
      class="blogsContainer"
    >
      <div
        data-selected="true"
        data-testid="blogCard"
      />
      <div
        data-selected="false"
        data-testid="blogCard"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Blog list snapshots should match blog list with no blogs 1`] = `
<DocumentFragment>
  <div
    class="blogListContainer"
  >
    <div
      class="searchBar"
    >
      <input
        class="search"
        placeholder="Search Blogs"
        type="text"
        value=""
      />
      <button
        class="btn pinkBtn"
      >
        NEW
      </button>
    </div>
    <div
      class="blogsContainer"
    >
      <p
        class="noBlogs"
      >
        There is no blogs available
      </p>
    </div>
  </div>
</DocumentFragment>
`;
